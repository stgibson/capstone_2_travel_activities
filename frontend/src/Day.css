.Day {
  margin-bottom: 20px;
}