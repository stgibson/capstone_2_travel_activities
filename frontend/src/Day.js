import React, { useState, useEffect } from "react";
import { useSelector, useDispatch, shallowEqual } from "react-redux";
import Activity from "./Activity";
import BaseForm from "./BaseForm";
import {
  addActivityToDayInAPI,
  removeActivityFromDayInAPI
} from "./actions/plans";
import "./Day.css";

/**
 * Component for displaying a day in a travel plan
 * @param {Object} param0 
 * @returns JSX code for rendering a day and its activities
 */
const Day = ({ day, planId }) => {
  const [inputChoices, setInputChoices] = useState([]);
  const [inputs, setInputs] = useState([]);
  const token = useSelector(store => store.token);
  const favoriteActivities =
    useSelector(store => store.activities, shallowEqual);
  const dispatch = useDispatch();

  /**
   * If activity with id data.activity isn't already in day, add to day in API
   * and redux store by calling dispatch
   * @param {Object{string}} data 
   */
  const handleAddActivityToPlan = data => {
    const activityId = Number.parseInt(data.activity);
    if (!day.activities.find(activity => activity.id === activityId)) {
      dispatch(addActivityToDayInAPI(
        planId,
        day.number,
        activityId,
        token
      ));
    }
  };

  /**
   * Removes activity with button clicked from day in API and redux store by
   * calling dispatch
   * @param {Object} evt 
   */
  const handleRemoveActivityFromPlan = evt => {
    const { id } = evt.target.parentElement;
    const activityId = Number.parseInt(id);
    dispatch(removeActivityFromDayInAPI(
      planId,
      day.number,
      activityId,
      token
    ));
  };

  // get array of user's favorite activities
  useEffect(() => {
    const favoriteActivitiesArr = Object.keys(favoriteActivities)
      .map(activityId => favoriteActivities[activityId]);
    setInputChoices(favoriteActivitiesArr);
  }, [favoriteActivities, setInputChoices]);

  // generate inputs for form to add activity
  useEffect(() => {
    setInputs([
      {
        name: "activity",
        type: "select",
        choices: inputChoices
      }
    ]);
  }, [inputChoices])

  return (
    <div id={ day.number } className="Day">
      <h3>{ day.number }</h3>
      {
        day.activities.map(activity => (
          <div key={ activity.id }>
            <Activity
              activity={ activity }
              btnText="Remove"
              btnVariant="warning"
              btnCallback={ evt => handleRemoveActivityFromPlan(evt) }
            />
          </div>
        ))
      }
      <BaseForm
        inputs={ inputs }
        btnText="Add Activity"
        submitCallback={ handleAddActivityToPlan }
      />
    </div>
  );
};

export default Day;