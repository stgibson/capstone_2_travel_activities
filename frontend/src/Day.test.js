import React from "react";
import { MemoryRouter } from "react-router-dom";
import { createStore } from "redux";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import Day from "./Day";
import rootReducer from "./reducers/root";

const store = createStore(rootReducer);

const activities = [
  { id: 1, name: "Test Activity 1" },
  { id: 2, name: "Test Activity 2" }
];
const day = { number: 1, activities };

it("renders without crashing", () => {
  render(
    <Provider store={ store }>
      <MemoryRouter><Day day={ day } planId={ 1 } /></MemoryRouter>
    </Provider>
  );
});

it("matches snapshot", () => {
  const { asFragment } = render(
    <Provider store={ store }>
      <MemoryRouter><Day day={ day } planId={ 1 } /></MemoryRouter>
    </Provider>
  );

  expect(asFragment()).toMatchSnapshot();
});
