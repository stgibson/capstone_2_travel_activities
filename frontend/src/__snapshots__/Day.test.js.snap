// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Day"
    id="1"
  >
    <h3>
      1
    </h3>
    <div>
      <div
        class="Activity"
        id="1"
      >
        <a
          class="Activity-link"
          href="/activities/1"
        >
          Test Activity 1
        </a>
        <button
          class="btn btn-warning btn-sm"
          type="button"
        >
          Remove
        </button>
      </div>
    </div>
    <div>
      <div
        class="Activity"
        id="2"
      >
        <a
          class="Activity-link"
          href="/activities/2"
        >
          Test Activity 2
        </a>
        <button
          class="btn btn-warning btn-sm"
          type="button"
        >
          Remove
        </button>
      </div>
    </div>
    <form
      class=""
    >
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="activity"
        />
        <select
          class="form-control"
          id="activity"
        />
      </div>
    </form>
    <button
      class="btn btn-primary"
      type="button"
    >
      Add Activity
    </button>
  </div>
</DocumentFragment>
`;
